freud/util.pyx

```diff
@@ -195,7 +195,7 @@ cdef class _Compute(object):
         return repr(self)
 
 
-def _convert_array(array, shape=None, dtype=np.float32):
+def _convert_array(array, shape=None, dtype=np.float32, inplace=False):
```

Comment:
I think we should make this function capable of requiring a copy. Right now box wrapping (and possibly a few other places) contain extra internal logic to either make an extra copy or not, and adding in place operations makes that logic more complicated than simply adding a `.copy()` to the output of this function. I would suggest adding a new argument `copy="optimize"` that has the following options:

- `default` (better name welcome): This would match the current behavior: copy if needed. Errors if any of the shape/dimensionality checks fail, or the `np.require` call fails. Whether or not a copy is made is entirely determined by whether or not `np.require` makes a copy internally.
- `inplace`: Never makes a copy. Errors if any of the shape/dimensionality checks fail. Also errors if the input array does not have the correct flags. This can either be implemented by checking the flags directly (which this PR currently does), or by calling `np.require` and raising an Exception if the output is a copy. This method would also remove the initial call to `np.asarray` and replace it with a type check
- `copy`: Always returns a copy of the input array. Errors if any of the shape/dimensionality checks fail. Replace the call to `np.require` with a call to `np.array(array, dtype=dtype, order='C')`. 

We will want to make sure that copies are minimized. That also means that the initial `np.asarray` call should probably pass the `dtype` and `order` arguments through in the `default` and `copy` cases, if we're going to retain the call. Moreover, it's redundant to call both `np.asarray` and `np.require` the way that we do now, so we can simplify that if we figure out the exact right numpy function that makes copies in the situations we want.